lisa/_assets/kmodules/lisa/fs.c

```diff
@@ -0,0 +1,755 @@
+		return -ENOMEM;
+
+	init_feature_param_entry(entry, cfg, &lisa_features_param);
+
+	/* set_features: enable a feature - RW. */
+	if (!lisa_fs_create_single(parent, "set_features",
+			      NULL,
+			      &lisa_param_feature_fops,
+			      S_IFREG | S_IRUGO | S_IWUGO, entry)) {
+		free_feature_param_entry(entry);
+		return -ENOMEM;
+	}
+
+	/* available_features: list available features - RO. */
+	if (!lisa_fs_create_single(parent, "available_features",
+			      NULL,
+			      &lisa_available_features_fops,
+			      S_IFREG | S_IRUGO, &lisa_features_param))
+		return -ENOMEM;
+
+	/* activate: activate the selected (and configured) features - RW. */
+	if (!lisa_fs_create_single(parent, "activate",
+			      NULL,
+			      &lisa_activate_fops,
+			      S_IFREG | S_IRUGO | S_IWUGO, cfg))
+		return -ENOMEM;
+
+	/* configs: Dir containing configurations, only setup at the top level. */
+	if (is_top_level) {
+		if (!lisa_fs_create_single(parent, "configs",
+				&lisa_fs_dir_inode_operations,
+				&simple_dir_operations,
+				S_IFDIR | S_IRUGO, NULL))
+			return -ENOMEM;
+	}
+
+	/* Create a dir for features having parameters. */
+	for_each_feature(feature) {
+		struct feature_param *param, **pparam;
+		struct dentry *dentry;
+
+		if (!feature->params)
+			continue;
+
+		dentry = lisa_fs_create_single(parent, feature->name,
+					       &simple_dir_inode_operations,
+					       &simple_dir_operations,
+					       S_IFDIR | S_IRUGO, cfg);
+		if (!dentry) {
+			pr_err("Failed to initialize feature's (%s) root node\n",
+			       feature->name);
+			return -ENOMEM;
+		}
+
+		for_each_feature_param(param, pparam, feature) {
+			entry = allocate_feature_param_entry();
+			if (!entry)
+				return -ENOMEM;
+
+			init_feature_param_entry(entry, cfg, param);
+
+			if (!lisa_fs_create_single(dentry, param->name, NULL,
+						   &lisa_param_feature_fops,
+						   S_IFREG | S_IRUGO, entry)) {
+				free_feature_param_entry(entry);
+				return -ENOMEM;
+			}
+		}
+	}
+
+	return 0;
+}
+
+/////////////////////////////////////
+//           Super block
+/////////////////////////////////////
+
+static struct super_operations lisa_super_ops = {
+	.statfs = simple_statfs,
+};
+
+static int lisa_fs_fill_super(struct super_block *sb, struct fs_context *fc)
+{
+	struct lisa_cfg *cfg;
+	struct inode *root;
+	int ret = -ENOMEM;
+
+	sb->s_maxbytes = MAX_LFS_FILESIZE;
+	sb->s_blocksize = PAGE_SIZE;
+	sb->s_blocksize_bits = PAGE_SHIFT;
+	sb->s_magic = LISA_FS_SUPER_MAGIC;
+	sb->s_op = &lisa_super_ops;
+
+	root = lisa_fs_create_inode(sb, S_IFDIR | S_IRUGO);
+	if (!root)
+		return -ENOMEM;
+
+	root->i_op = &simple_dir_inode_operations;
+	root->i_fop = &simple_dir_operations;
+
+	sb->s_root = d_make_root(root);
+	if (!sb->s_root)
+		goto error1;
+
+	cfg = allocate_lisa_cfg("root");
+	if (!cfg)
+		goto error2;
+
+	mutex_lock(&interface_lock);
+
+	ret = lisa_fs_create_files(sb->s_root, true, cfg);
+	if (ret)
+		goto error3;
+
+	ret = init_lisa_cfg(cfg, &cfg_list, NULL);
```

Comment:
is this supposed to be 'global' config ? If so why not passing the s_root dentry ?

Reply:
Yes right